### PR TITLE
Refactor lookup of cartesian index and element data 

### PR DESCRIPTION
**Pull request by aritorto**
To search data via element Index and Cartesian Index, the classes LookUpData and LookUpCartesianData have been added. Subjected to OPM/opm-grid#670.

### CODE REVIEW

**Comment by aritorto** [comment]
jenkins build this serial please

**Comment by aritorto** [comment]
I see, then I delete the comment!

**Comment by aritorto** [comment]
jenkins build this opm-grid=670 serial please

**Comment by blattms** [comment]
Mmh, maybe you are not white-listed for this repository, yet. Let me try again.

**Comment by blattms** [comment]
jenkins build this opm-grid=670 serial please

File: ebos/eclgenericthresholdpressure.cc [hunk 1]
```diff
@@ -77,14 +79,9 @@ thresholdPressure(int elem1Idx, int elem2Idx) const
 
     // threshold pressure accross faults
     if (!thpresftValues_.empty()) {
-        int cartElem1Idx = cartMapper_.cartesianIndex(elem1Idx);
-        int cartElem2Idx = cartMapper_.cartesianIndex(elem2Idx);
-
-        assert(0 <= cartElem1Idx && static_cast<int>(cartElemFaultIdx_.size()) > cartElem1Idx);
-        assert(0 <= cartElem2Idx && static_cast<int>(cartElemFaultIdx_.size()) > cartElem2Idx);
```
Comment (blattms) [inline ebos/eclgenericthresholdpressure.cc:84]: This check is now done inside of LookUpCartesianData, right?

File: ebos/eclgenericthresholdpressure.cc [hunk 2]
```diff
@@ -182,11 +179,8 @@ applyExplicitThresholdPressures_()
             const auto& inside = intersection.inside();
             const auto& outside = intersection.outside();
 
-            unsigned insideElemIdx = elementMapper_.index(inside);
-            unsigned outsideElemIdx = elementMapper_.index(outside);
-
-            auto equilRegionInside = elemEquilRegion_[insideElemIdx];
-            auto equilRegionOutside = elemEquilRegion_[outsideElemIdx];
+            unsigned equilRegionInside = lookUpData_(inside, elemEquilRegion_);
+            unsigned equilRegionOutside = lookUpData_(outside, elemEquilRegion_);
```
Comment (blattms) [inline ebos/eclgenericthresholdpressure.cc:183]: I think auto should stay there. The entries are actually `unsigned short`.

**Review by blattms — CHANGES_REQUESTED** [review]
This looks already nice. :sunglasses: 

I just would like to see two changes. Hopefully one does fix the regression.

Hopefully unrelated:
I have no idea where thesee [additional warnings](https://ci.opm-project.org/job/opm-simulators-PR-builder/5157/gcc/folder.1573348989/) in opm-common from. I thought they might be due to OPM/opm-common#3585, yet I don't see them there. Maybe somebody else has some insight. (Here we are doing a serial build as oppossed to the other PR).

File: ebos/eclinterregflows.cc [hunk 3]
```diff
@@ -55,7 +55,7 @@ addConnection(const Cell& source,
     }
 
     if (! source.isInterior ||
-        (source.cartesianIndex > destination.cartesianIndex))
+        (source.activeIndex > destination.activeIndex))
```
Comment (blattms) [inline ebos/eclinterregflows.cc:58]: This one got lost in the review:
This change looks strange. Might this be the cause for the failing test?

File: ebos/eclgenericthresholdpressure.cc (hunk 1 above)
Comment (aritorto) [reply ebos/eclgenericthresholdpressure.cc:84]: Indeed, lines 323-324 (also lines 313-14) in LookUpData.hh

File: ebos/eclinterregflows.cc (hunk 3 above)
Comment (aritorto) [reply ebos/eclinterregflows.cc:58]: I don't remember why it got changed. I'll turn it into the old `cartesianIndex` then. Thanks!

**Comment by aritorto** [comment]
Thanks for the review. The changes can be found in the last commit.

**Comment by aritorto** [comment]
jenkins build this opm-grid=670 serial please

**Comment by goncalvesmachadoc** [comment]
jenkins build this opm-grid=670 serial please

**Comment by aritorto** [comment]
jenkins build this opm-grid=670 serial please

**Comment by blattms** [comment]
I will postpone merging this. I need to find out where the additional warnings come from.

**Comment by blattms** [comment]
benchmark opm-grid=670 please

**Comment by blattms** [comment]
Should have looked at this a long time ago. Sorry. Will try to investigate now where the additional warnings in opm-simulators come from.

**Comment by blattms** [comment]
In the meantime would you please rebase this on current master. Unfortunately, some of the touched files changed quite a bit.  My apologies.

**Comment by aritorto** [comment]
jenkins build this serial please

**Comment by aritorto** [comment]
jenkins build this serial please

**Comment by blattms** [comment]
Somehow the additional warnings vanished. :sunglasses:

**Comment by blattms** [comment]
jenkins build this please

**Comment by aritorto** [comment]
jenkins build this serial please

**Comment by blattms** [comment]
benchmark please

**Comment by blattms** [comment]
mmh, I did get an email with the benchmark results, but somehow they are not listed here. I just post it here manually then:
```

| Test    | Configuration                                     | Relative |
| ------- | ------------------------------------------------- | -------- |
| opm-git | OPM Benchmark: drogon - Threads: 1                | 1.003    |
| opm-git | OPM Benchmark: drogon - Threads: 8                | 0.99     |
| opm-git | OPM Benchmark: punqs3 - Threads: 1                | 1.011    |
| opm-git | OPM Benchmark: punqs3 - Threads: 8                | 1.006    |
| opm-git | OPM Benchmark: smeaheia - Threads: 1              | 1.001    |
| opm-git | OPM Benchmark: smeaheia - Threads: 8              | 1.007    |
| opm-git | OPM Benchmark: spe10_model_1 - Threads: 1         | 1.019    |
| opm-git | OPM Benchmark: spe10_model_1 - Threads: 8         | 0.996    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1        | 1.014    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8        | 1        |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1        | 0.993    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8        | 1.018    |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 1 | 0.991    |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 8 | 1.014    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *
```
And in it goes. Thanks a lot.
